### ISSUE TITLE
Failed Message details URL

### ISSUE THREAD

**Issue by Tsjunne** [Particular/ServicePulse#526]
It would be very nice to be able to predict the URL to a given failed message in ServicePulse.
I tried integrating a monitoring mail with a link to ServicePulse, but than I noticed that the ID in the URL (/#/failed-messages/message/{id}) is not the actual NServiceBus.MessageId.
It would be real nice to have a URL that can be generated in a mail or other medium without needing access to the internal ID in SC.

- **ServicePulse version:** 
1.9.1
- **ServiceControl version:** 
1.41.2

**Comment by indualagarsamy** [comment]
Thanks for raising the issue @Tsjunne. Can you have a look at this issue: https://github.com/Particular/ServicePulse/issues/331. I believe in version 1.8.0 we added the ability to bookmark a URL. Can you add more details as to the use case you're after?

**Comment by WilliamBZA** [comment]
Hi @Tsjunne 

The problem is that the `NServiceBus.MessageId` header is not unique within ServiceControl. Once a message ends up in ServiceControl as a failed message for the first time, the message Id might be unique (but not guaranteed to be). If that message is retried and ends up back in the error queue, it's definitely no longer unique - which is why we introduced the `FailedMessageId` we use to direct link to the message. That Id is unique to each message failure, or each time a message appears in ServiceControl.

If I understand what you're trying to achieve correctly, you want to be able to link to the LAST time that the message was attempted to be processed, in other words: if you link to the message ID, you would want to see the last instance of that message ID regardless of how many other times that message ID appears within ServiceControl. Does this sound correct?

**Comment by Tsjunne** [comment]
Hi @WilliamBZA 

That's correct, something like that would work. I understand that the NSB.MessageId isn't unique in ServiceControl. So either a search results page or the semantics that you suggest would do the trick for us. The main problem is that we don't have any correlation key coming from the logs that is usable to construct a URL to the failed message details in Pulse.
I could achieve the same thing, but it would involve querying the API to get the search result, parsing the JSON and extracting the failedMessageId. But this would involve a lot of custom scripting in our monitoring tools or an elaborate proxy.

**Comment by WilliamBZA** [comment]
Got it @Tsjunne.  I can see the value in a URL route like that. We don't have anything to match that at the moment, so let's leave this issue open and see if we can get something for you when one of the maintainers has time to implement it.

**Comment by Tsjunne** [comment]
I managed to solve this for our case in a structural way by setting up an extra endpoint as part of our monitoring infrastructure that subscribes on the ServiceControl events, as illustrated in the [docs](https://docs.particular.net/servicecontrol/contracts). These events contain the internal FailedMessageId that can be used to point to the relevant URL in ServicePulse. 
So now, in stead of monitoring the NServicebus log events, I now just monitor the logs of this new service.

**Comment by Tsjunne** [comment]
None the less, it would still be nice to have a way to filter on MessageId in the ServicePulse UI

**Comment by WilliamBZA** [comment]
Glad to hear it @Tsjunne, rest assured I will get a dedicated URL route in for you. I'll ping here as soon as I do!

**Comment by VincentVrijburg** [comment]
@WilliamBZA I believe it is still not possible to link to ServicePulse based on the messageid that failed, correct?

I have a similar use case as the OP where I would like to link to ServicePulse from an alert email using the messageid. It seems that the latest version of SC/SP still uses internal ids and the UI does not allow for filtering based on messageid.

**Comment by TravisNickels** [comment]
@VincentVrijburg

> I believe it is still not possible to link to ServicePulse based on the messageid that failed, correct?

Yes, that's correct there is no API path to view a failed Message based on the messageId.

However, as a workaround, the failedMessageId that is used in ServicePulse to view the failed message is deterministic based on the messageId and the name of the endpoint that processed the message.  The [DeterministicGuid](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Infrastructure/DeterministicGuid.cs) is not a public class, but can be copied into you code to generate the failedMessageID.  When calling the MakeId() method the first parameter is the messageId and the second parameter is the endpointName like in this [example](https://github.com/Particular/ServiceControl/blob/master/src/ServiceControl/Infrastructure/TransportMessageExtensions.cs#L44).
